pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -345,63 +403,37 @@ func (h *handler) OnRemove(key string, obj runtime.Object) (runtime.Object, erro
 
 	if _, nodeDrainExcluded := machine.Annotations[capi.ExcludeNodeDrainingAnnotation]; !removed ||
 		(!nodeDrainExcluded && !drainingSucceededCondition.IsTrue(machine) && drainingSucceededCondition.GetReason(machine) != capi.DrainingFailedReason) {
-		if err = h.dynamic.EnqueueAfter(obj.GetObjectKind().GroupVersionKind(), objMeta.GetNamespace(), objMeta.GetName(), 5*time.Second); err != nil {
+		if err = h.dynamic.EnqueueAfter(obj.GetObjectKind().GroupVersionKind(), infraObj.meta.GetNamespace(), infraObj.meta.GetName(), 5*time.Second); err != nil {
 			return obj, err
 		}
 		return obj, generic.ErrSkip
 	}
 
-	return h.doRemove(obj, objMeta)
+	return h.doRemove(infraObj)
 }
 
-func (h *handler) doRemove(obj runtime.Object, objMeta metav1.Object) (runtime.Object, error) {
-	d, err := data.Convert(obj.DeepCopyObject())
-	if err != nil {
-		return nil, err
-	}
-
-	infraName := objMeta.GetName()
-	if !d.Bool("status", "jobComplete") && d.String("status", "failureReason") == "" {
-		return obj, fmt.Errorf("cannot delete machine %s because create job has not finished", infraName)
-	}
-
-	obj, err = h.run(obj, objMeta, false, 3)
+func (h *handler) doRemove(infraObj *infraObject) (runtime.Object, error) {
+	obj, err := h.run(infraObj, false, 3)
 	if err != nil {
 		return nil, err
 	}
-	job, err := h.jobs.Get(objMeta.GetNamespace(), getJobName(infraName))
-	if err != nil {
-		return nil, err
-	}
-
-	if !job.Status.CompletionTime.IsZero() || shouldForceRemoval(job, d) {
```

Comment:
For reviewers: This is no longer needed because the check was moved up.

Reply:
Is it OK we've essentially flipped the order? Previously it was `h.run` first and then cleanup of the objects and now it's the other way around.